ouroboros-consensus/src/Ouroboros/Consensus/Ledger/SupportsMempool.hs

```diff
@@ -33,31 +35,34 @@ type family ApplyTxErr blk :: Type
 
 class ( UpdateLedger blk
       , NoThunks (GenTx blk)
+      , NoThunks (Validated (GenTx blk))
       , NoThunks (Ticked (LedgerState blk))
       , Show (GenTx blk)
+      , Show (Validated (GenTx blk))
       , Show (ApplyTxErr blk)
       ) => LedgerSupportsMempool blk where
 
   -- | Check whether the internal invariants of the transaction hold.
   txInvariant :: GenTx blk -> Bool
   txInvariant = const True
 
-  -- | Apply transaction we have not previously seen before
+  -- | Apply an unvalidated transaction
   applyTx :: LedgerConfig blk
           -> SlotNo -- ^ Slot number of the block containing the tx
           -> GenTx blk
           -> TickedLedgerState blk
-          -> Except (ApplyTxErr blk) (TickedLedgerState blk)
+          -> Except (ApplyTxErr blk) (TickedLedgerState blk, Validated (GenTx blk))
```

Comment:
This is the root change.

`applyTx` is now essentially the "smart constructor" for validated transactions. And it's also at the heart of the mempool.

Together, those two facts lead to the most of the remaining diff in this PR. (Another "lemma" of sorts to keep in mind is that blocks contain validated transactions, so that's what is now passed to the `forge*` functions etc.) (And another: we don't send validated things across to our peers.)